src/inputfetcher.h

```diff
@@ -0,0 +1,203 @@
+                }
+                if (auto coin{m_db->GetCoin(outpoint)}; coin) {
+                    m_coins[thread_index].emplace_back(outpoint, std::move(*coin));
+                } else {
+                    // Missing an input. This block will fail validation.
+                    // Skip remaining inputs.
+                    m_input_counter.store(m_inputs.size(), std::memory_order_relaxed);
+                    return;
+                }
+            }
+        } catch (const std::runtime_error& e) {
+            // Database error. This will be handled later in validation.
+            // Skip remaining inputs.
+            LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Warning, "InputFetcher failed to fetch input: %s.\n", e.what());
+            m_input_counter.store(m_inputs.size(), std::memory_order_relaxed);
+        }
+    }
+
+    void OnCompletion() noexcept
+    {
+        // At this point all threads are done writing to m_coins and reading from m_cache,
+        // so we can safely read from m_coins and write to m_cache.
+        for (auto& coins : m_coins) {
+            for (auto&& [outpoint, coin] : coins) {
+                m_cache->EmplaceCoinInternalDANGER(std::move(outpoint), std::move(coin));
+            }
+            coins.clear();
+        }
+    }
+
+public:
+    explicit InputFetcher(size_t worker_thread_count) noexcept
+        : m_complete_barrier{static_cast<int32_t>(worker_thread_count + 1), OnCompletionWrapper{this}}
+    {
+        if (worker_thread_count == 0) {
+            return;
+        }
+        for (size_t n{0}; n < worker_thread_count; ++n) {
+            m_coins.emplace_back();
+            m_worker_threads.emplace_back(std::bind(&InputFetcher::ThreadLoop, this, n));
+        }
+        // One more coins vector for the main thread
+        m_coins.emplace_back();
+    }
+
+    //! Fetch all block inputs from db, and insert into cache.
+    void FetchInputs(CCoinsViewCache& cache, const CCoinsView& db, const CBlock& block) noexcept
+    {
+        if (block.vtx.size() <= 1 || m_worker_threads.size() == 0) {
+            return;
+        }
+
+        m_db = &db;
+        m_cache = &cache;
+        m_block = &block;
+        m_txids.clear();
+        m_inputs.clear();
+
+        // Loop through the inputs of the block and add them to the queue.
+        // Construct the set of txids to filter.
+        for (size_t i{1}; i < block.vtx.size(); ++i) {
+            for (size_t j{0}; j < block.vtx[i]->vin.size(); ++j) {
+                m_inputs.emplace_back(i, j);
+            }
+            m_txids.emplace(block.vtx[i]->GetHash());
+        }
+
+        // Set the input counter and wake threads.
+        m_input_counter.store(0, std::memory_order_relaxed);
+        m_start_semaphore.release(m_worker_threads.size());
+
+        // Have the main thread work too before we wait for other threads
+        Work(m_worker_threads.size());
+        m_complete_barrier.arrive_and_wait();
```

Comment:
What's the reason for not doing the completion here instead of the `OnCompletion` workaround?
```suggestion
        m_complete_barrier.arrive_and_wait();
        for (auto& coins : m_coins) {
            for (auto&& [outpoint, coin] : coins) {
                m_cache->EmplaceCoinInternalDANGER(std::move(outpoint), std::move(coin));
            }
            coins.clear();
        }
```